<html><head></head><body>
<div id="sbo-rt-content"><section aria-labelledby="ch23" epub:type="chapter" role="doc-chapter">
<span aria-label="441" epub:type="pagebreak" id="pg_441" role="doc-pagebreak"/>
<hgroup>
<h2 class="CHAPTER" id="ch23">
<span class="CN"><span class="SANS_Futura_Std_Bold_Condensed_B_11">23</span></span>
<span class="CT"><span class="SANS_Dogma_OT_Bold_B_11">ERROR HANDLING WITH EXCEPTIONS</span></span>
</h2>
</hgroup>
<figure class="opener"><img alt="" class="opener" height="380" src="../images/opener.jpg" width="380"/>
</figure>
<p class="INTRO">Applications don’t always function the way you want. For example, a file may not upload because of a network error, or data from a user or web API may be malformed in some way. In this chapter, you’ll learn to use <i>exceptions</i> to anticipate these sorts of problems and recover from them, so your application doesn’t always crash when something goes wrong. You’ll work with PHP’s generic <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> class, along with other, more specialized exception classes built into the language. You’ll also see how to design your own custom exception classes, as well as how to design your applications to safely handle any and all exceptions that may arise.</p>
<section aria-labelledby="sec1" epub:type="division">
<span aria-label="442" epub:type="pagebreak" id="pg_442" role="doc-pagebreak"/>
<h3 class="H1" id="sec1"><span id="toc-link_308"/><span class="SANS_Futura_Std_Bold_B_11">The Basics of Exceptions</span></h3>
<p class="TNI1"><i>Exceptions</i> are classes that provide a sophisticated and customizable approach to handling and recovering from anticipated, problematic circumstances in OOP. They differ from <i>errors</i>, which arise from circumstances or events that can’t be recovered from, such as the computer system running out of memory or a class declaration attempting to use a constant that can’t be found. PHP has a built-in <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> class for handling generic problems, along with other more specialized exception classes that cater to particular types of errors. You can also develop your own custom exception classes.</p>
<p class="TX">Exception-based software design allows you to write code in the most natural sequence, assuming it will all work fine, and then to separately write code to capture and address any typical problems that may occur. This involves writing tests into the methods of a class that generate exception objects and disrupt the flow of program control whenever an unusual or invalid situation occurs, such as providing invalid arguments for a constructor or setter method. Thanks to these tests, code appearing later in a method can be written with the safe assumption that if execution gets that far, the exception-throwing conditions haven’t occurred, and the code is working as it should.</p>
<p class="TX">Central to exception-based application programming are <span class="SANS_TheSansMonoCd_W5Regular_11">throw</span> and <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statements. A method uses a <span class="SANS_TheSansMonoCd_W5Regular_11">throw</span> statement to create an exception object when a problem occurs. This is called <i>throwing an exception</i>. The <span class="SANS_TheSansMonoCd_W5Regular_11">throw</span> statement halts the execution of the method and disrupts the flow of the program. By itself, throwing an exception can lead to a fatal error, unless you <i>catch</i> the exception with a <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement. The <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement features code that’s intended to be executed when an exception is thrown; such code may allow the application to recover from the issue, or if non-recoverable, then the problem can be logged and execution ended gracefully.</p>
<p class="TX">In this section, we’ll explore the basics of throwing and catching exceptions. We’ll also look at <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> statements, pieces of code that are executed at the end of a process, regardless of whether an exception has been thrown.</p>
<section aria-labelledby="sec2" epub:type="division">
<h4 class="H2" id="sec2"><span id="toc-link_309"/><span class="SANS_Futura_Std_Bold_Condensed_Oblique_11">Throwing an Exception</span></h4>
<p class="TNI1">First, we’ll consider how to throw an uncaught exception to cause a fatal error, halting an application. We’ll examine the common use case of throwing an exception when an invalid argument is provided to the setter method of a class. We’ll create a variation of the <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span> and <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> classes from <span class="Xref"><a href="chapter19.xhtml">Chapter 19</a></span>, adding exception-based validation behavior in the <span class="SANS_TheSansMonoCd_W5Regular_11">setCalories()</span> method of the <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> class. An exception will be thrown if a negative value is provided as the number of calories for a new <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object. The project we’ll create is illustrated in the UML class diagram in <a href="#fig23-1">Figure 23-1</a>.</p>
<span aria-label="443" epub:type="pagebreak" id="pg_443" role="doc-pagebreak"/>
<figure class="IMG"><a id="fig23-1"/><img alt="" class="img100" height="739" src="../images/figure23-1.jpg" width="1426"/>
<figcaption><p class="CAP"><span class="SANS_Futura_Std_Book_Oblique_11">Figure 23-1: A diagram showing an exception thrown by the</span> <span class="SANS_TheSansMonoCd_W5Regular_Italic_11">Dessert</span> <span class="SANS_Futura_Std_Book_Oblique_11">class</span></p></figcaption>
</figure>
<p class="TX">Recall that <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> is a subclass of <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span>, with its own <span class="SANS_TheSansMonoCd_W5Regular_11">__toString()</span> method and a <span class="SANS_TheSansMonoCd_W5Regular_11">calories</span> property. The diagram indicates that an invalid <span class="SANS_TheSansMonoCd_W5Regular_11">calories</span> value will throw an exception.</p>
<p class="TX">First, we’ll declare the <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span> class. Create a new project with <i>src/Food.php</i> containing the code in <a href="#lis23-1">Listing 23-1</a>.</p>
<span id="lis23-1"/>
<pre><code>&lt;?php&#13;
namespace Mattsmithdev;&#13;
&#13;
class Food&#13;
{&#13;
    protected string $name;&#13;
&#13;
    public function __construct(string $name)&#13;
    {&#13;
        $this-&gt;name = $name;&#13;
    }&#13;
&#13;
    public function __toString(): string&#13;
    {&#13;
        return "(FOOD) $this-&gt;name";&#13;
    }&#13;
}</code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-1: The</span> <span class="TheSansMonoCd_W5Regular_Italic_11">Food</span> <span class="SANS_Futura_Std_Book_Oblique_11">superclass</span></p>
<p class="TX">We assign the <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span> class to the <span class="SANS_TheSansMonoCd_W5Regular_11">Mattsmithdev</span> namespace and give it a <span class="SANS_TheSansMonoCd_W5Regular_11">name</span> property with <span class="SANS_TheSansMonoCd_W5Regular_11">protected</span> visibility so that all subclasses can directly access it. The class has a straightforward constructor to initialize <span class="SANS_TheSansMonoCd_W5Regular_11">name</span> when each new object is created, and a <span class="SANS_TheSansMonoCd_W5Regular_11">__toString()</span> method to return a string in the form <span class="SANS_TheSansMonoCd_W5Regular_11">"(FOOD)</span> <span class="SANS_TheSansMonoCd_W5Regular_Italic_11">name</span><span class="SANS_TheSansMonoCd_W5Regular_11">"</span>.</p>
<p class="TX">Let’s now declare the <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> subclass of <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span>. Create <i>src/Dessert.php</i> and enter the contents of <a href="#lis23-2">Listing 23-2</a>.</p>
<span id="lis23-2"/>
<pre><code><span aria-label="444" epub:type="pagebreak" id="pg_444" role="doc-pagebreak"/>&lt;?php&#13;
namespace Mattsmithdev;&#13;
&#13;
class Dessert extends Food&#13;
{&#13;
    private int $calories;&#13;
&#13;
    public function __construct(string $name, int $calories)&#13;
    {&#13;
        parent::__construct($name);&#13;
      <span aria-label="annotation1" class="Code_CodeAnnotation">❶</span> $this-&gt;setCalories($calories);&#13;
    }&#13;
    public function getCalories(): int&#13;
    {&#13;
        return $this-&gt;calories;&#13;
    }&#13;
&#13;
    public function setCalories(int $calories)&#13;
    {&#13;
      <span aria-label="annotation2" class="Code_CodeAnnotation">❷</span> if ($calories &lt; 0) {&#13;
            throw new \Exception(&#13;
                'attempting to set calories to a negative value');&#13;
        }&#13;
&#13;
      <span aria-label="annotation3" class="Code_CodeAnnotation">❸</span> $this-&gt;calories = $calories;&#13;
    }&#13;
&#13;
    public function __toString(): string&#13;
    {&#13;
        return "I am a Dessert containing $this-&gt;calories!";&#13;
    }&#13;
}</code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-2: The</span> <span class="TheSansMonoCd_W5Regular_Italic_11">Dessert</span> <span class="SANS_Futura_Std_Book_Oblique_11">class, which throws an exception if the</span> <span class="TheSansMonoCd_W5Regular_Italic_11">calories</span> <span class="SANS_Futura_Std_Book_Oblique_11">value is invalid</span></p>
<p class="TX">The <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> class has a <span class="SANS_TheSansMonoCd_W5Regular_11">calories</span> property assigned in the constructor via the <span class="SANS_TheSansMonoCd_W5Regular_11">setCalories()</span> method <span aria-label="annotation1" class="CodeAnnotation">❶</span>. This way, we reserve any validation logic for the setter method itself, so every new <span class="SANS_TheSansMonoCd_W5Regular_11">calories</span> value will be vetted, regardless of whether it’s provided at the time of object construction or via the setter at a later point.</p>
<p class="TX">Within <span class="SANS_TheSansMonoCd_W5Regular_11">setCalories()</span>, we perform the validation with an <span class="SANS_TheSansMonoCd_W5Regular_11">if</span> statement <span aria-label="annotation2" class="CodeAnnotation">❷</span>. If the provided integer argument <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span> is less than <span class="SANS_TheSansMonoCd_W5Regular_11">0</span>, we throw a new <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> object, with the message <span class="SANS_TheSansMonoCd_W5Regular_11">'attempting to set calories to a negative value'</span>. If the <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span> argument is <span class="SANS_TheSansMonoCd_W5Regular_11">0</span> or more and no exception is thrown, the code continues by storing the provided value in the <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object’s <span class="SANS_TheSansMonoCd_W5Regular_11">calories</span> property <span aria-label="annotation3" class="CodeAnnotation">❸</span>.</p>
<p class="TX">Notice the syntax for throwing the exception. We begin with the <span class="SANS_TheSansMonoCd_W5Regular_11">throw</span> keyword, which tells PHP to disrupt the flow of the program if the <span class="SANS_TheSansMonoCd_W5Regular_11">if</span> statement is <span class="SANS_TheSansMonoCd_W5Regular_11">true</span>. Then we use the <span class="SANS_TheSansMonoCd_W5Regular_11">new</span> keyword to create a new object of the <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> class, passing the error message we want to display as an argument. We have to prefix the <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> class with a backslash (<span class="SANS_TheSansMonoCd_W5Regular_11">\</span>) because it’s part of PHP’s root namespace, whereas <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> is part of the <span class="SANS_TheSansMonoCd_W5Regular_11">Mattsmithdev</span> <span aria-label="445" epub:type="pagebreak" id="pg_445" role="doc-pagebreak"/>namespace. Without the backslash, <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> would be assumed to be in the <span class="SANS_TheSansMonoCd_W5Regular_11">Mattsmithdev</span> namespace as well.</p>
<p class="TX">We next need to write a <i>composer.json</i> file to autoload our classes. Create this file as shown in <a href="#lis23-3">Listing 23-3</a>.</p>
<span id="lis23-3"/>
<pre><code>{&#13;
    "autoload": {&#13;
        "psr-4": {&#13;
            "Mattsmithdev\\": "src"&#13;
        }&#13;
    }&#13;
}</code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-3: The</span> <span class="SANS_Futura_Std_Book_11">composer.json</span> <span class="SANS_Futura_Std_Book_Oblique_11">file for autoloading</span></p>
<p class="BodyContinued">Once you have this file, generate the autoloader scripts by entering <span class="SANS_TheSansMonoCd_W7Bold_11">composer dump-autoload</span> at the command line.</p>
<p class="TX">Now let’s write an index script to attempt to create a <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span> and a <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object. Create <i>public/index.php</i> to match <a href="#lis23-4">Listing 23-4</a>.</p>
<span id="lis23-4"/>
<pre><code>&lt;?php&#13;
require_once __DIR__ . '/../vendor/autoload.php';&#13;
&#13;
use Mattsmithdev\Food;&#13;
use Mattsmithdev\Dessert;&#13;
&#13;
$f1 = new Food('apple');&#13;
print $f1 . PHP_EOL;&#13;
&#13;
$f2 = new Dessert('strawberry cheesecake', -1);&#13;
print $f2;</code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-4: Attempting to create an invalid</span> <span class="TheSansMonoCd_W5Regular_Italic_11">Dessert</span> <span class="SANS_Futura_Std_Book_Oblique_11">object in</span> <span class="SANS_Futura_Std_Book_11">index.php</span></p>
<p class="TX">We read and execute the autoloader and provide <span class="SANS_TheSansMonoCd_W5Regular_11">use</span> statements for the two classes we need. Then we create and print a <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span> object and a <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object, passing an invalid argument of <span class="SANS_TheSansMonoCd_W5Regular_11">-1</span> for the latter’s <span class="SANS_TheSansMonoCd_W5Regular_11">calories</span> property. Here’s the result of running this index script at the command line:</p>
<pre><code>$ <b>php public/index.php</b>&#13;
(FOOD) apple&#13;
&#13;
Fatal error: Uncaught Exception: attempting to set calories to a negative&#13;
value in /Users/matt/src/Dessert.php:23&#13;
Stack trace:&#13;
#0 /Users/matt/src/Dessert.php(12): Mattsmithdev\Dessert-&gt;setCalories(-1)&#13;
#1 /Users/matt/public/index.php(10): Mattsmithdev\Dessert-&gt;__construct&#13;
('strawberry chee...', -1)&#13;
#2 {main}&#13;
  thrown in /Users/matt/src/Dessert.php on line 23</code></pre>
<p class="TX">The first line of output shows that the <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span> object was successfully created and printed out, but then we get a fatal error due to the exception <span aria-label="446" epub:type="pagebreak" id="pg_446" role="doc-pagebreak"/>thrown by the negative calorie value. The exception is said to be <i>uncaught</i>, since we didn’t write any code telling PHP what to do if an exception is thrown. As a result, the application has simply stopped running and has printed the error message we provided, followed by a <i>stack trace</i>, a report that steps through the code to show the cause of the exception.</p>
<p class="TX">The stack trace tells us the following:</p>
<ul class="ul">
<li class="ListBullet"><span class="SANS_TheSansMonoCd_W5Regular_11">#0</span> shows that the exception was thrown when <span class="SANS_TheSansMonoCd_W5Regular_11">setCalories()</span> was passed <span class="SANS_TheSansMonoCd_W5Regular_11">-1</span> as an argument at line 12 of the <i>src/Dessert.php</i> file.</li>
<li class="ListBullet"><span class="SANS_TheSansMonoCd_W5Regular_11">#1</span> shows that <span class="SANS_TheSansMonoCd_W5Regular_11">setCalories()</span> was called when the <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> class’s constructor method was invoked with the arguments <span class="SANS_TheSansMonoCd_W5Regular_11">('strawberry chee...', -1)</span> (the food-name string has been shortened).</li>
<li class="ListBullet"><span class="SANS_TheSansMonoCd_W5Regular_11">#2</span> reports that the exception-throwing code is line 23 in <i>src/Dessert.php</i>.</li>
</ul>
<p class="TX">Notice that the output ends with the stack trace, meaning the index script wasn’t able to get to the point of printing out the <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object. The uncaught exception halted the flow of the program, so the final line of the index script didn’t execute.</p>
</section>
<section aria-labelledby="sec3" epub:type="division">
<h4 class="H2" id="sec3"><span id="toc-link_310"/><span class="SANS_Futura_Std_Bold_Condensed_Oblique_11">Catching an Exception</span></h4>
<p class="TNI1">To avoid a fatal error and safely manage exceptions, we need to <i>catch</i> the exceptions by writing a <span class="SANS_TheSansMonoCd_W5Regular_11">try...catch</span> statement in our index script. The <span class="SANS_TheSansMonoCd_W5Regular_11">try</span> portion indicates what we want to do under normal circumstances, and the <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> portion indicates what to do when an exception is thrown.</p>
<p class="TX">By catching exceptions, we prevent application users from seeing fatal errors and the resulting stack traces. Besides being embarrassing and not user-friendly, printing out a stack trace “leaks” information about the structure of the web application code (in the preceding example, for instance, we leaked the folder name <i>src</i> and the class filename <i>Dessert.php</i>). While stack traces aren’t serious security vulnerability issues, any information leaked like this might be helpful to an attacker and so should be prevented where possible. Catching exceptions lets us decide what to do with the exception data, as well as what the user will see when a problem arises.</p>
<blockquote>
<p class="Note"><span class="SANS_Dogma_OT_Bold_B_15">NOTE</span></p>
</blockquote>
<p class="NOTE-TXT"><i>In <a href="chapter24.xhtml">Chapter 24</a>, we’ll explore logging, which allows useful debugging data such as stack traces to be stored for developers and site administrators to access, while not publishing such information to any public website visitor or software client.</i></p>
<p class="TX">To catch the exception raised when a negative calorie value is given, update the <i>public/index.php</i> script to match <a href="#lis23-5">Listing 23-5</a>.</p>
<span id="lis23-5"/>
<pre><code><span class="TheSansMonoCd_W5Regular_Grey_11">&lt;?php</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">require_once __DIR__ . '/../vendor/autoload.php';</span>&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">use Mattsmithdev \Food;</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">use Mattsmithdev \Dessert;</span>&#13;
&#13;
<span aria-label="447" epub:type="pagebreak" id="pg_447" role="doc-pagebreak"/><span aria-label="annotation1" class="codeannotated_CodeAnnotation">❶</span> try {&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    $f1 = new Food('apple');</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print $f1 . PHP_EOL;</span>&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    $f2 = new Dessert('strawberry cheesecake', -1);</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print $f2;</span>&#13;
<span aria-label="annotation2" class="codeannotated_CodeAnnotation">❷</span>} catch (\Exception $e) {&#13;
    print '(caught!) - an exception occurred!' . PHP_EOL;&#13;
  <span aria-label="annotation3" class="Code_CodeAnnotation">❸</span> print $e-&gt;getMessage();&#13;
}</code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-5: Adding a</span> <span class="TheSansMonoCd_W5Regular_Italic_11">try...catch</span> <span class="SANS_Futura_Std_Book_Oblique_11">statement to</span> <span class="SANS_Futura_Std_Book_11">index.php</span></p>
<p class="TX">The old code creating and printing <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span> and <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> objects is now inside a <span class="SANS_TheSansMonoCd_W5Regular_11">try</span> block <span aria-label="annotation1" class="CodeAnnotation">❶</span>. If any exception occurs during this sequence, PHP checks the class of the exception against the class(es) specified in the <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> block that follows <span aria-label="annotation2" class="CodeAnnotation">❷</span>. If the class matches, the <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> block is executed. In this case, the <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement is for objects of the <span class="SANS_TheSansMonoCd_W5Regular_11">\Exception</span> class, as specified in the parentheses after the <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> keyword. The variable <span class="SANS_TheSansMonoCd_W5Regular_11">$e</span>, also in the parentheses, becomes a reference to the <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> object that has been caught.</p>
<p class="TX">In the <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> block, we print out the message <span class="SANS_TheSansMonoCd_W5Regular_11">'(caught!) - an exception occurred!'</span> followed by a line break. Then we print the message inside the <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> object via its public <span class="SANS_TheSansMonoCd_W5Regular_11">getMessage()</span> method <span aria-label="annotation3" class="CodeAnnotation">❸</span>. This is the <span class="SANS_TheSansMonoCd_W5Regular_11">'attempting to set calories to a negative value'</span> message we defined earlier.</p>
<p class="TX">Now that we’ve added code catching the exception, try running the index script again at the command line. You should see the following:</p>
<pre><code>$ <b>php public/index.php</b>&#13;
(FOOD) apple&#13;
(caught!) - an exception occurred!&#13;
attempting to set calories to a negative value</code></pre>
<p class="TX">Again, the <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span> object has been successfully created and printed. Next, we see the message printed from inside our <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement, followed by the message from the <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> object itself. In this example, having caught the exception, we’re still printing out a message for the user, but we’ve controlled the information that’s displayed. No stack trace is leaking information now that we’re handling the exception with a <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement.</p>
</section>
<section aria-labelledby="sec4" epub:type="division">
<h4 class="H2" id="sec4"><span id="toc-link_311"/><span class="SANS_Futura_Std_Bold_Condensed_Oblique_11">Ending with a finally Statement</span></h4>
<p class="TNI1">A <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> statement is a block of code that gets executed regardless of whether an exception has been thrown. It’s written after the <span class="SANS_TheSansMonoCd_W5Regular_11">try</span> and <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statements and typically includes <i>housekeeping code</i>, code that gracefully ends any processes that have been started. For example, you might use a <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> statement to ensure that any file streams or database connections are closed.</p>
<p class="TX">Let’s add a <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> statement to our index script to gracefully close the application every time it runs, even if an exception has been thrown. Modify <i>public/index.php</i> to match <a href="#lis23-6">Listing 23-6</a>.</p>
<span id="lis23-6"/>
<pre><code><span aria-label="448" epub:type="pagebreak" id="pg_448" role="doc-pagebreak"/><span class="TheSansMonoCd_W5Regular_Grey_11">&lt;?php</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">require_once __DIR__ . '/../vendor/autoload.php';</span>&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">try {</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    $f1 = new Food('apple');</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print $f1 . PHP_EOL;</span>&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    $f2 = new Dessert('strawberry cheesecake', -1);</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print $f2;</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">} catch (\Exception $e) {</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print '(caught!) - an exception occurred!' . PHP_EOL;</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print $e-&gt;getMessage();</span>&#13;
} finally {&#13;
    print PHP_EOL  . '(finally) -- Application finished --';&#13;
}</code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-6: Adding a</span> <span class="TheSansMonoCd_W5Regular_Italic_11">finally</span> <span class="SANS_Futura_Std_Book_Oblique_11">statement to</span> <span class="SANS_Futura_Std_Book_11">index.php</span></p>
<p class="TX">We declare a <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> block that prints a simple message after either the <span class="SANS_TheSansMonoCd_W5Regular_11">try</span> or <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> block concludes. Here’s the result of running this updated index script:</p>
<pre><code>$ <b>php public/index.php</b>&#13;
(FOOD) apple&#13;
(caught!) - an exception occurred!&#13;
attempting to set calories to a negative value&#13;
(finally) -- Application finished --</code></pre>
<p class="BodyContinued">The message from the <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> block prints at the end of the output, after displaying the message from the exception.</p>
<p class="TX">To make sure the <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> statement executes even when the <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement doesn’t, let’s update our script so that the <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object is given a valid number of calories, meaning no exception will be thrown. Modify the instantiation of the <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object in <i>public/index.php</i> as shown in <a href="#lis23-7">Listing 23-7</a>.</p>
<span id="lis23-7"/>
<pre><code><span class="TheSansMonoCd_W5Regular_Italic_11">--snip--</span>&#13;
    $f2 = new Dessert('strawberry cheesecake', <b>500</b>);&#13;
<span class="TheSansMonoCd_W5Regular_Italic_11">--snip--</span></code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-7: Creating a valid</span> <span class="TheSansMonoCd_W5Regular_Italic_11">Dessert</span> <span class="SANS_Futura_Std_Book_Oblique_11">object in</span> <span class="SANS_Futura_Std_Book_11">index.php</span></p>
<p class="TX">When you run the index script now, you should see the <span class="SANS_TheSansMonoCd_W5Regular_11">Food</span> and <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object messages and then the <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> message:</p>
<pre><code>$ <b>php public/index.php</b>&#13;
(FOOD) apple&#13;
I am a Dessert containing 500 calories!&#13;
(finally) -- Application finished --</code></pre>
<p class="BodyContinued">The output confirms that a <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object was successfully created without throwing an exception, and that the <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> block was still executed regardless.</p>
</section>
</section>
<section aria-labelledby="sec5" epub:type="division">
<span aria-label="449" epub:type="pagebreak" id="pg_449" role="doc-pagebreak"/>
<h3 class="H1" id="sec5"><span id="toc-link_312"/><span class="SANS_Futura_Std_Bold_B_11">Using Multiple Exception Classes</span></h3>
<p class="TNI1">In addition to PHP’s root <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> class, several other classes of exception are available as part of the Standard PHP Library (SPL), such as the <span class="SANS_TheSansMonoCd_W5Regular_11">InvalidArgumentException</span> class. These other exception classes are all connected hierarchically to <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> as its subclasses, subclasses of its subclasses, and so on. You can also create your own custom exception classes that are subclasses of one of these built-in exception classes.</p>
<p class="TX">At first glance, it may seem unnecessary to have subclasses of <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span>, since we could create basic <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> objects with custom messages for each situation throwing an exception. However, by writing code that throws objects of different <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> subclasses, you can include several <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statements, one for each <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> subclass, allowing you to respond differently to each type of exception.</p>
<p class="TX">For example, you could write multiple validation checks into a setter method and throw a certain class of exception depending on which validation check fails. Then you could write a separate <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement for each of the exception classes, so each type of exception generates a customized response. You would then typically end with a <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement for generic <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> objects, allowing you to catch any exceptions you didn’t already account for. We’ll look at how this works in the following sections.</p>
<section aria-labelledby="sec6" epub:type="division">
<h4 class="H2" id="sec6"><span id="toc-link_313"/><span class="SANS_Futura_Std_Bold_Condensed_Oblique_11">Other Built-in Exception Classes</span></h4>
<p class="TNI1">Let’s use another built-in PHP exception class in conjunction with the root <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> class. We’ll update our <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> class’s <span class="SANS_TheSansMonoCd_W5Regular_11">setCalories()</span> method to throw one of two exception class objects as part of the validation of the received <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span> argument. Our validation tests are as follows:</p>
<ul class="ul">
<li class="ListBullet">If <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span> is less than <span class="SANS_TheSansMonoCd_W5Regular_11">0</span>, throw an <span class="SANS_TheSansMonoCd_W5Regular_11">\InvalidArgumentException</span> object because desserts can’t have negative calories.</li>
<li class="ListBullet">If <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span> is greater than <span class="SANS_TheSansMonoCd_W5Regular_11">5000</span>, throw a general <span class="SANS_TheSansMonoCd_W5Regular_11">\Exception</span> object because that’s <i>way too many</i> calories for one dessert.</li>
</ul>
<p class="TX">Update the <span class="SANS_TheSansMonoCd_W5Regular_11">setCalories()</span> method in <i>src/Dessert.php</i> to match <a href="#lis23-8">Listing 23-8</a>.</p>
<span id="lis23-8"/>
<pre><code><span class="TheSansMonoCd_W5Regular_Grey_11">public function setCalories(int $calories): void</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">{</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    if ($calories &lt; 0) {</span>&#13;
        throw new \InvalidArgumentException(&#13;
            'attempting to set calories to a negative value');&#13;
&#13;
    }&#13;
&#13;
    if ($calories &gt; 5000) {&#13;
        throw new \Exception('too many calories for one dessert!');&#13;
    }&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    $this-&gt;calories = $calories;</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">}</span></code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-8: Updating the</span> <span class="TheSansMonoCd_W5Regular_Italic_11">setCalories()</span> <span class="SANS_Futura_Std_Book_Oblique_11">method to throw different classes of exception</span></p>
<p class="TX"><span aria-label="450" epub:type="pagebreak" id="pg_450" role="doc-pagebreak"/>First, we change the class of exception thrown when the argument received is negative to <span class="SANS_TheSansMonoCd_W5Regular_11">\InvalidArgumentException</span>. Once again, note the forward slash before the class name, indicating that this class is declared in the root PHP namespace. Then we add a second validation test: when the number of calories is greater than <span class="SANS_TheSansMonoCd_W5Regular_11">5000</span>, an object of the general <span class="SANS_TheSansMonoCd_W5Regular_11">\Exception</span> class will be thrown. If execution of the code gets past these two <span class="SANS_TheSansMonoCd_W5Regular_11">if</span> statements without throwing any exceptions, we store the provided value in the object’s <span class="SANS_TheSansMonoCd_W5Regular_11">calories</span> property as before.</p>
<p class="TX">Next, we need to update the index script. We’ll write multiple <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statements to handle each class of exception object appropriately. Then we’ll try different values for the <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object’s <span class="SANS_TheSansMonoCd_W5Regular_11">calories</span> property to test the validation logic. Edit <i>public/index.php</i> to match <a href="#lis23-9">Listing 23-9</a>.</p>
<span id="lis23-9"/>
<pre><code><span class="TheSansMonoCd_W5Regular_Grey_11">&lt;?php</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">require_once __DIR__ . '/../vendor/autoload.php';</span>&#13;
&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">use Mattsmithdev\Food;</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">use Mattsmithdev\Dessert;</span>&#13;
&#13;
$calories = -1; // Negative invalid argument&#13;
$calories = 6000; // General exception&#13;
$calories = 500; // Valid&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">try {</span>&#13;
    $f2 = new Dessert('strawberry cheesecake', $calories);&#13;
    print $f2;&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">}</span> <span aria-label="annotation1" class="Code_CodeAnnotation">❶</span> catch (\InvalidArgumentException $e) {&#13;
    print '(caught!) - an Invalid Argument Exception occurred!' . PHP_EOL;&#13;
    print $e-&gt;getMessage();&#13;
} <span aria-label="annotation2" class="Code_CodeAnnotation">❷</span> catch (\Exception $e) {&#13;
    print '(caught!) - a general Exception occurred!' . PHP_EOL;&#13;
    print $e-&gt;getMessage();&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">} finally {</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print PHP_EOL . '(finally) -- Application finished --';</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">}</span></code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-9: Multiple</span> <span class="TheSansMonoCd_W5Regular_Italic_11">catch</span> <span class="SANS_Futura_Std_Book_Oblique_11">statements in the</span> <span class="SANS_Futura_Std_Book_11">public/index.php</span> <span class="SANS_Futura_Std_Book_Oblique_11">script</span></p>
<p class="TX">We start with three assignment statements for different values of the <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span> variable. To thoroughly test the script, comment out all but one of these assignment statements, choosing a different one each time. In the <span class="SANS_TheSansMonoCd_W5Regular_11">try</span> block, we create a new <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object, providing the <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span> variable as an argument. Then we create two <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statements, one for the <span class="SANS_TheSansMonoCd_W5Regular_11">InvalidArgumentException</span> class <span aria-label="annotation1" class="CodeAnnotation">❶</span> and the other for the general <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> class <span aria-label="annotation2" class="CodeAnnotation">❷</span>. Each prints a different message, along with the message attached to the exception object itself, retrieved with <span class="SANS_TheSansMonoCd_W5Regular_11">$e-&gt;getMessage()</span>.</p>
<p class="TX"><span aria-label="451" epub:type="pagebreak" id="pg_451" role="doc-pagebreak"/><a href="#tab23-1">Table 23-1</a> shows the outputs for the three values of <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span>, demonstrating that our exception-based logic is working as expected.</p>
<p class="TT"><a id="tab23-1"/><span class="Heavy"><span class="SANS_Futura_Std_Heavy_11">Table 23-1:</span></span> <span class="SANS_Futura_Std_Book_11">Outputs for Calorie Values</span></p>
<table class="Basic-Table">
<thead>
<tr>
<th class="TCH" scope="col"><p class="TCH1"><span class="SANS_Futura_Std_Heavy_11">Value of</span> <span class="TheSansMonoCd_W7Bold_11-1">$calories</span></p></th>
<th class="TCH" scope="col"><p class="TCH1"><span class="SANS_Futura_Std_Heavy_11">Program output</span></p></th>
</tr>
</thead>
<tbody>
<tr>
<td class="TBF"><p class="TB_Grey"><span class="SANS_TheSansMonoCd_W5Regular_11">-1</span></p></td>
<td class="TBF"><p class="TB_Grey"><span class="SANS_TheSansMonoCd_W5Regular_11">(caught!) - an Invalid Argument Exception occurred! attempting to set calories to a negative value (finally) -- Application finished --</span></p></td>
</tr>
<tr>
<td class="TB"><p class="TB1"><span class="SANS_TheSansMonoCd_W5Regular_11">6000</span></p></td>
<td class="TB"><p class="TB1"><span class="SANS_TheSansMonoCd_W5Regular_11">(caught!) - a general Exception occurred! too many calories for one dessert! (finally) -- Application finished --</span></p></td>
</tr>
<tr>
<td class="TBL"><p class="TB_Grey"><span class="SANS_TheSansMonoCd_W5Regular_11">500</span></p></td>
<td class="TBL"><p class="TB_Grey"><span class="SANS_TheSansMonoCd_W5Regular_11">I am a Dessert containing 500 calories! (finally) -- Application finished --</span></p></td>
</tr>
</tbody>
</table>
<p class="TX">Notice that the values of <span class="SANS_TheSansMonoCd_W5Regular_11">-1</span> and <span class="SANS_TheSansMonoCd_W5Regular_11">6000</span> each trigger their own class of exception, while <span class="SANS_TheSansMonoCd_W5Regular_11">500</span> allows the <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object to be successfully created and printed.</p>
</section>
<section aria-labelledby="sec7" epub:type="division">
<h4 class="H2" id="sec7"><span id="toc-link_314"/><span class="SANS_Futura_Std_Bold_Condensed_Oblique_11">Custom Exception Classes</span></h4>
<p class="TNI1">PHP gives you the flexibility to write your own custom exception classes, provided they’re subclasses of <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> or one of the other built-in PHP exception classes. Also, many third-party libraries come with their own custom exception classes designed specifically for the methods in that library. Whether you’re writing your own or using someone else’s, custom exception classes give you even more freedom to structure your code to respond differently to a variety of anticipated problems.</p>
<p class="TX">Let’s add a custom exception class to our <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> project: <span class="SANS_TheSansMonoCd_W5Regular_11">Mattsmithdev\NegativeCaloriesException</span>. We’ll update the project to throw an exception object of this class instead of the <span class="SANS_TheSansMonoCd_W5Regular_11">InvalidArgumentException</span> class. <a href="#fig23-2">Figure 23-2</a> shows the two classes of exception that our <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> objects can throw. Notice that the <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> class falls within the <span class="SANS_TheSansMonoCd_W5Regular_11">Mattsmithdev</span> namespace, while the <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> class is outside, since it’s in the root PHP namespace.</p>
<span aria-label="452" epub:type="pagebreak" id="pg_452" role="doc-pagebreak"/>
<figure class="IMG"><a id="fig23-2"/><img alt="" class="img100" height="895" src="../images/figure23-2.jpg" width="1365"/>
<figcaption><p class="CAP"><span class="SANS_Futura_Std_Book_Oblique_11">Figure 23-2: The two classes of exception that a</span> <span class="SANS_TheSansMonoCd_W5Regular_Italic_11">Dessert</span> <span class="SANS_Futura_Std_Book_Oblique_11">object can throw</span></p></figcaption>
</figure>
<p class="TX">First, create a new class in <i>src/NegativeCaloriesException.php</i> containing the code in <a href="#lis23-10">Listing 23-10</a>.</p>
<span id="lis23-10"/>
<pre><code>&lt;?php&#13;
namespace Mattsmithdev;&#13;
&#13;
class NegativeCaloriesException extends \Exception&#13;
{&#13;
}</code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-10: The custom</span> <span class="TheSansMonoCd_W5Regular_Italic_11">NegativeCaloriesException</span> <span class="SANS_Futura_Std_Book_Oblique_11">class</span></p>
<p class="TX">We declare <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> as a subclass of the root <span class="SANS_TheSansMonoCd_W5Regular_11">\Exception</span> class. It contains no methods. Since it doesn’t have its own constructor method, it will inherit the constructor from its <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> superclass, allowing it to take in a message for display.</p>
<p class="TX">Let’s now update our <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> class’s <span class="SANS_TheSansMonoCd_W5Regular_11">setCalories()</span> method to throw a <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> object when the provided calorie value is negative. As in the previous example, we’ll throw a general <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> object when the provided value is greater than <span class="SANS_TheSansMonoCd_W5Regular_11">5000</span>. Update the <span class="SANS_TheSansMonoCd_W5Regular_11">setCalories()</span> method in <i>src/Dessert.php</i> to match <a href="#lis23-11">Listing 23-11</a>.</p>
<span id="lis23-11"/>
<pre><code><span class="TheSansMonoCd_W5Regular_Grey_11">public function setCalories(int $calories)</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">{</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    if ($calories &lt; 0) {</span>&#13;
        throw new NegativeCaloriesException(&#13;
            'attempting to set calories to a negative value');&#13;
    }&#13;
&#13;
<span aria-label="453" epub:type="pagebreak" id="pg_453" role="doc-pagebreak"/><span class="TheSansMonoCd_W5Regular_Grey_11">    if ($calories &gt; 5000) {</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">        throw new \Exception('too many calories for one dessert!');</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    }</span>&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    $this-&gt;calories = $calories;</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">}</span></code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-11: Throwing a custom exception in the</span> <span class="TheSansMonoCd_W5Regular_Italic_11">setCalories()</span> <span class="SANS_Futura_Std_Book_Oblique_11">method</span></p>
<p class="TX">We change the class of exception thrown when the argument received is negative to an object of the <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> class. Since this new class is in the same namespace as our <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> class, we don’t write a backslash before the class identifier.</p>
<p class="TX">Next, we need to update the <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statements in our index script to handle the new custom exception class. Modify <i>public/index.php</i> as shown in <a href="#lis23-12">Listing 23-12</a>.</p>
<span id="lis23-12"/>
<pre><code><span class="TheSansMonoCd_W5Regular_Grey_11">&lt;?php</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">require_once __DIR__ . '/../vendor/autoload.php';</span>&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">use Mattsmithdev\Dessert;</span>&#13;
use Mattsmithdev\NegativeCaloriesException;&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">$calories = -1; // Negative invalid argument</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">$calories = 6000; // General exception</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">$calories = 500; // Valid</span>&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">try {</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    $f2 = new Dessert('strawberry cheesecake', $calories);</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print $f2;</span>&#13;
} catch (NegativeCaloriesException) {&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print '(caught!) - a Negative Calories Value Exception occurred!' . PHP_EOL;</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print $e-&gt;getMessage();</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">} catch (\Exception $e) {</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print '(caught!) - a general Exception occurred! ' . PHP_EOL;</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print $e-&gt;getMessage();</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">} finally {</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">    print PHP_EOL . '(finally) -- Application finished --';</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">}</span></code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-12: Catching custom exception objects in the</span> <span class="SANS_Futura_Std_Book_11">index.php</span> <span class="SANS_Futura_Std_Book_Oblique_11">script</span></p>
<p class="TX">We add a <span class="SANS_TheSansMonoCd_W5Regular_11">use</span> statement so we can reference the <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> class without the <span class="SANS_TheSansMonoCd_W5Regular_11">Mattsmithdev</span> namespace prefix. Then we create a <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement for exceptions of this class, printing an appropriate message. Here’s the output you should get if you try to create a new <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object with <span class="SANS_TheSansMonoCd_W5Regular_11">-1</span> calories, confirming that a <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> is thrown:</p>
<pre><code>$ <b>php public/index.php</b>&#13;
(caught!) - a Negative Calories Value Exception occurred!&#13;
attempting to set calories to a negative value&#13;
(finally) -- Application finished --</code></pre>
<p class="TX"><span aria-label="454" epub:type="pagebreak" id="pg_454" role="doc-pagebreak"/>Testing for a negative value is a simple example, but it serves to illustrate how straightforward it is to create custom subclasses of <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span>, allowing you to write different logic to address different anticipated problems at runtime.</p>
</section>
</section>
<section aria-labelledby="sec8" epub:type="division">
<h3 class="H1" id="sec8"><span id="toc-link_315"/><span class="SANS_Futura_Std_Bold_B_11">Call-Stack Bubbling</span></h3>
<p class="TNI1">If an exception occurs in a block of code and isn’t caught by that code block, it will <i>bubble up</i> the call stack to the code that invoked the code block. If not caught there, the exception will continue to bubble up through successively higher levels of code until either it’s caught and handled or the top of the call stack is reached. If the exception isn’t caught in the top-level block of code, a fatal error will result, as we saw in this chapter’s first example. For this reason, it’s a good idea to include some code at the top level of an application to catch any exceptions that may have bubbled all the way to the top of the call stack.</p>
<p class="TX">As you’ve seen, the flow of control for a typical object-oriented PHP web application begins with the index script, which creates an object of the <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class and invokes its <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method. This in turn triggers the creation of other objects and the invocation of other methods. Some of this activity might throw exceptions. You can try to catch all those exceptions within the <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class, but any uncaught exceptions will ultimately bubble up to the index script and should be caught there to avoid a fatal error.</p>
<p class="TX">To demonstrate how this works, we’ll update our <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> project, taking the code that used to be in the index script and encapsulating it in an <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class. This class will be responsible for catching any <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> objects thrown during the creation of <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> objects, but we’ll allow other miscellaneous exceptions to bubble up to the top of the call stack. Then we’ll catch those in the top-level index script.</p>
<p class="TX">First, let’s update our index script to create an <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> object and invoke its <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method. We’ll wrap that code in a <span class="SANS_TheSansMonoCd_W5Regular_11">try...catch</span> statement to handle any uncaught exceptions that bubble up, and we’ll include a <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> statement to gracefully close the application. We’ll also clearly indicate that the messages being printed by the <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> and <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> blocks are coming from this index script by prefixing them with <span class="SANS_TheSansMonoCd_W5Regular_11">(index.php)</span>. Modify <i>public/index.php</i> to match <a href="#lis23-13">Listing 23-13</a>.</p>
<span id="lis23-13"/>
<pre><code><span class="TheSansMonoCd_W5Regular_Grey_11">&lt;?php</span>&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">require_once __DIR__ . '/../vendor/autoload.php';</span>&#13;
&#13;
use Mattsmithdev\Application;&#13;
&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">try {</span>&#13;
    $app = new Application();&#13;
    $app-&gt;run();&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">} catch (\Exception $e) {</span>&#13;
    print '(index.php) Exception caught!';&#13;
}<b> </b>finally {&#13;
<span aria-label="455" epub:type="pagebreak" id="pg_455" role="doc-pagebreak"/>    print PHP_EOL;&#13;
    print '(index.php) finally -- Application finished --';&#13;
<span class="TheSansMonoCd_W5Regular_Grey_11">}</span></code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-13: The simplified</span> <span class="SANS_Futura_Std_Book_11">index.php</span> <span class="SANS_Futura_Std_Book_Oblique_11">script creating an</span> <span class="TheSansMonoCd_W5Regular_Italic_11">Application</span> <span class="SANS_Futura_Std_Book_Oblique_11">object</span></p>
<p class="TX">Inside the <span class="SANS_TheSansMonoCd_W5Regular_11">try</span> block, we create an <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> object, storing a reference to the new object in the <span class="SANS_TheSansMonoCd_W5Regular_11">$app</span> variable, and invoke its <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method. If any uncaught <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> objects bubble up from the <span class="SANS_TheSansMonoCd_W5Regular_11">try</span> block statements, we use a <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> block to handle them and print a message. Since every exception object is an instance of the <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> class (either directly or as a subclass), this <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> statement acts as a catchall for any possible exception object that was thrown during program execution but wasn’t caught elsewhere in the code. We also add a <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> block that will print a final message regardless of whether any exceptions were thrown.</p>
<p class="TX">Now let’s write the <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class. Create a new file named <i>src/Application.php</i> to match <a href="#lis23-14">Listing 23-14</a>.</p>
<span id="lis23-14"/>
<pre><code>&lt;?php&#13;
namespace Mattsmithdev;&#13;
&#13;
class Application&#13;
{&#13;
    public function run(): void&#13;
    {&#13;
        $calories = -1; // Negative invalid argument&#13;
        $calories = 6000; // General exception&#13;
        $calories = 500; // Valid&#13;
&#13;
        try {&#13;
            $f2 = new Dessert('strawberry cheesecake', $calories);&#13;
            print $f2;&#13;
        } catch (NegativeCaloriesException $e) {&#13;
            print&#13;
'(Application-&gt;run) - Negative Calories Value Exception caught!';&#13;
        }&#13;
    }&#13;
}</code></pre>
<p class="CodeListingCaption"><span class="SANS_Futura_Std_Book_Oblique_11">Listing 23-14: The</span> <span class="TheSansMonoCd_W5Regular_Italic_11">Application</span> <span class="SANS_Futura_Std_Book_Oblique_11">class</span></p>
<p class="TX">We declare the <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class with a <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method containing many of the statements from the older index script. As before, we include three assignment statements for the <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span> variable that you can selectively comment out to test the project. Then we create and print a new <span class="SANS_TheSansMonoCd_W5Regular_11">Dessert</span> object in a <span class="SANS_TheSansMonoCd_W5Regular_11">try</span> block and use a <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> block to handle <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> objects. <a href="#tab23-2">Table 23-2</a> shows the results of running the application with the different possible values of <span class="SANS_TheSansMonoCd_W5Regular_11">$calories</span>.</p>
<span aria-label="456" epub:type="pagebreak" id="pg_456" role="doc-pagebreak"/>
<p class="TT"><a id="tab23-2"/><span class="Heavy"><span class="SANS_Futura_Std_Heavy_11">Table 23-2:</span></span> <span class="SANS_Futura_Std_Book_11">Catching Exceptions with Call-Stack Bubbling</span></p>
<table class="Basic-Table">
<thead>
<tr>
<th class="TCH" scope="col"><p class="TCH1"><span class="SANS_Futura_Std_Heavy_11">Value of</span> <span class="SANS_TheSansMonoCd_W7Bold_11">$calories</span></p></th>
<th class="TCH" scope="col"><p class="TCH1"><span class="SANS_Futura_Std_Heavy_11">Program output</span></p></th>
</tr>
</thead>
<tbody>
<tr>
<td class="TBF"><p class="TB_Grey"><span class="SANS_TheSansMonoCd_W5Regular_11">500</span></p></td>
<td class="TBF"><p class="TB_Grey"><span class="SANS_TheSansMonoCd_W5Regular_11">I am a Dessert containing 500 calories! (index.php) finally -- Application finished --</span></p></td>
</tr>
<tr>
<td class="TB"><p class="TB1"><span class="SANS_TheSansMonoCd_W5Regular_11">-1</span></p></td>
<td class="TB"><p class="TB1"><span class="SANS_TheSansMonoCd_W5Regular_11">(Application-&gt;run) - Negative Calories Value Exception caught! (index.php) finally -- Application finished --</span></p></td>
</tr>
<tr>
<td class="TBL"><p class="TB_Grey"><span class="SANS_TheSansMonoCd_W5Regular_11">6000</span></p></td>
<td class="TBL"><p class="TB_Grey"><span class="SANS_TheSansMonoCd_W5Regular_11">(index.php) Exception caught! (index.php) finally -- Application finished --</span></p></td>
</tr>
</tbody>
</table>
<p class="TX">When a valid value of <span class="SANS_TheSansMonoCd_W5Regular_11">500</span> is used, the object properties are printed out. When the value is <span class="SANS_TheSansMonoCd_W5Regular_11">-1</span>, the <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> is caught inside the <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method of the <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class. When the too-high value of <span class="SANS_TheSansMonoCd_W5Regular_11">6000</span> is used, the <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method of the <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class fails to catch the general <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> that’s thrown, since the method is watching only for <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span> objects. As a result, the exception bubbles up to the index script, where it hits the general <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> block. In all cases, the output ends with the message from the <span class="SANS_TheSansMonoCd_W5Regular_11">finally</span> block in the index script.</p>
<p class="TX">Adding a general <span class="SANS_TheSansMonoCd_W5Regular_11">try...catch</span> statement to an index script ensures that any bubbled-up uncaught exceptions will be addressed, meaning the application will avoid runtime errors relating to exceptions. Meanwhile, the code for handling more-specific exceptions, such as our custom <span class="SANS_TheSansMonoCd_W5Regular_11">NegativeCaloriesException</span>, is located at a lower level of the application code, which keeps the index script simple and well organized.</p>
</section>
<section aria-labelledby="sec9" epub:type="division">
<h3 class="H1" id="sec9"><span id="toc-link_316"/><span class="SANS_Futura_Std_Bold_B_11">Summary</span></h3>
<p class="TNI1">This chapter introduced how to work with exceptions. You learned how to create exceptions when anticipated problematic situations occur by writing <span class="SANS_TheSansMonoCd_W5Regular_11">throw</span> statements, and how to manage exceptions with <span class="SANS_TheSansMonoCd_W5Regular_11">try...catch...finally</span> structures. All exceptions are instances of PHP’s top-level <span class="SANS_TheSansMonoCd_W5Regular_11">\Exception</span> class, but we discussed how to refine the program logic by using provided exception subclasses such as <span class="SANS_TheSansMonoCd_W5Regular_11">InvalidArgumentException</span> or by declaring custom exception subclasses.</p>
<p class="TX">We also explored a general application architecture that exploits the bubbling up of uncaught exceptions. Specific anticipated exceptions can be caught within class methods, while any remaining exceptions can be caught in the index script at the top level of the application.</p>
</section>
<section aria-labelledby="sec10" epub:type="division">
<h3 class="H1" id="sec10"><span id="toc-link_317"/><span class="SANS_Futura_Std_Bold_B_11">Exercises</span></h3>
<p class="ListNumber">1.   Create a new project and implement a simple <span class="SANS_TheSansMonoCd_W5Regular_11">Product</span> class with private <span class="SANS_TheSansMonoCd_W5Regular_11">name</span> and <span class="SANS_TheSansMonoCd_W5Regular_11">price</span> properties, public accessor methods for each property, and a constructor that takes in new values for each property and sets them using the setter methods. Add validation as follows:</p>
<p class="ListLetterSub"><span aria-label="457" epub:type="pagebreak" id="pg_457" role="doc-pagebreak"/>a.   If a negative value for <span class="SANS_TheSansMonoCd_W5Regular_11">price</span> is received, an <span class="SANS_TheSansMonoCd_W5Regular_11">InvalidArgumentException</span> is thrown.</p>
<p class="ListLetterSub">b.   If a <span class="SANS_TheSansMonoCd_W5Regular_11">price</span> greater than <span class="SANS_TheSansMonoCd_W5Regular_11">1000000</span> is received, a general <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> is thrown.</p>
<p class="ListLetterSub">c.   If an empty string is provided for the <span class="SANS_TheSansMonoCd_W5Regular_11">name</span> property, an <span class="SANS_TheSansMonoCd_W5Regular_11">InvalidArgumentException</span> is thrown.</p>
<p class="ListBody">Create a <i>composer.json</i> file and an index script to attempt to create a <span class="SANS_TheSansMonoCd_W5Regular_11">Product</span> object with valid and invalid names and prices. Then wrap your index code with <span class="SANS_TheSansMonoCd_W5Regular_11">try...catch</span> statements, so you can handle the various exceptions your code throws.</p>
<p class="ListNumber">2.   Make a copy of your Exercise 1 solution and introduce an <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class similar to the one from <a href="#lis23-14">Listing 23-14</a>. Refactor your code as follows:</p>
<p class="ListLetterSub">a.   Create the <span class="SANS_TheSansMonoCd_W5Regular_11">Product</span> object in the <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method of your <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class. Catch <span class="SANS_TheSansMonoCd_W5Regular_11">InvalidArgumentException</span> objects and print an appropriate message as part of the <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method as well.</p>
<p class="ListLetterSub">b.   In the index script, create an <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> object and invoke its <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method. Catch any bubbled-up general <span class="SANS_TheSansMonoCd_W5Regular_11">Exception</span> objects and print an appropriate message.</p>
<p class="ListNumber">3.   Make a copy of your solution for Exercise 2 and introduce a custom exception named <span class="SANS_TheSansMonoCd_W5Regular_11">EmptyStringException</span> that’s thrown by the <span class="SANS_TheSansMonoCd_W5Regular_11">setName()</span> method. Add an appropriate <span class="SANS_TheSansMonoCd_W5Regular_11">catch</span> block to catch and process this exception in the <span class="SANS_TheSansMonoCd_W5Regular_11">run()</span> method of the <span class="SANS_TheSansMonoCd_W5Regular_11">Application</span> class.</p>
</section>
</section>
</div></body></html>